File: sources/java/_12_matrix_add.java
// task 12 matrix_add — expected output: 999000000
// build: javac _12_matrix_add.java    run: java _12_matrix_add
// build (graalvm native-image): native-image -O2 _12_matrix_add    run: ./_12_matrix_add
// Flat long[n*n] arrays with index i*n+j.

public class _12_matrix_add {
    public static void main(String[] args) {
        final int n = 1000;
        long[] a = new long[n * n];
        long[] b = new long[n * n];
        long[] c = new long[n * n];

        for (int i = 0; i < n; i++) {
            int row = i * n;
            for (int j = 0; j < n; j++) {
                a[row + j] = i + j;
                b[row + j] = i - j;
            }
        }
        for (int i = 0; i < n; i++) {
            int row = i * n;
            for (int j = 0; j < n; j++) {
                c[row + j] = a[row + j] + b[row + j];
            }
        }

        long total = 0;
        for (int idx = 0; idx < n * n; idx++) {
            total += c[idx];
        }
        System.out.println(total);
    }
}

File: sources/fortran/12_matrix_add.f90
! task 12 matrix_add — expected output: 999000000
! build: gfortran -O3 -o prog 12_matrix_add.f90 (flang -O3 -o prog 12_matrix_add.f90)    run: ./prog
!
! Flat arrays with the index i*n + j + 1, because Fortran subscripts are
! 1-based and one contiguous array is what the memory bandwidth is about.

program main
  implicit none
  integer, parameter :: n = 1000
  integer(kind=8), allocatable :: a(:), b(:), c(:)
  integer(kind=8) :: total, i, j, idx

  allocate(a(n*n), b(n*n), c(n*n))
  do j = 0_8, int(n, kind=8) - 1_8
    do i = 0_8, int(n, kind=8) - 1_8
      a(i*n + j + 1_8) = i + j
      b(i*n + j + 1_8) = i - j
    end do
  end do

  do idx = 1_8, int(n, kind=8)*int(n, kind=8)
    c(idx) = a(idx) + b(idx)
  end do

  total = 0_8
  do idx = 1_8, int(n, kind=8)*int(n, kind=8)
    total = total + c(idx)
  end do

  write(*,'(i0)') total
end program main
